HAR-3682 Ylläpidon muut työt

Jarzka opened the pull request.

Jarzka committed:
Lisää uudet taulut urakka_laskentakohde ja yllapito_toteuma

Jarzka committed:
Luo pohjaa tiemerkinnän muiden töiden lomakkeelle

Jarzka committed:
Lisää pvm ylläpidon toteumatauluun

Jarzka committed:
Luo pohja ylläpidon toteumien palvelulle

Jarzka committed:
Ota uusi palvelu mukaan järjestelmään

Jarzka committed:
Käytä alaviivaa uuden ylläpidon palvelutiedoston nimessä

Jarzka committed:
Lisää kyselyt & testidataa toteumille

Jarzka committed:
Päivitä migraatio, vaihda sarakkeiden järjestys mielekkäämpään

Jarzka committed:
Lisää valinnat muut työt -näkymään

Jarzka committed:
Rakenna tukea hakea työt gridiin

Jarzka committed:
Poista turha declare

Jarzka committed:
Nimeä ToidenHaku -> TyotHaettu

Jarzka committed:
Merge branch 'develop' into HAR-3682_yllapidon_muut_tyot

# Conflicts:
#	tietokanta/src/main/resources/db/migration/V1_461__.sql

Jarzka committed:
Merge branch 'develop' into HAR-3682_yllapidon_muut_tyot

# Conflicts:
#	tietokanta/src/main/resources/db/migration/V1_461__.sql

Jarzka committed:
Käytä eventin constructoria, älä tyyppiä

Jarzka committed:
Hae ylläpiton muut työt oikein gridiin

Jarzka committed:
Tee alustava tuki hakea työ

Jarzka committed:
Rakenna muiden töiden lomaketta

Jarzka committed:
Lisää tuki hakea työn perustiedot lomakkeelle

Jarzka committed:
Mahdollista uudet lomakkeen aloitus

Jarzka committed:
Säädä lomakkeen kenttien järjestys parempaan

Jarzka committed:
Aseta pakolliset kentät

Jarzka committed:
Luo palvelu toteuman tallennukseen ja päivitykseen

Jarzka committed:
Käytä Toteuma-nimitystä Työn sijaan

Jarzka committed:
Korjaa buildi

Jarzka committed:
Lue toteumat uudesta avaimesta oikein

Jarzka committed:
Lue valittu toteuma uudesta avaimesta oikein

Jarzka committed:
Lue valittu toteuma uudesta avaimesta oikein

Jarzka committed:
Korjaa typo SQL-kyselyssä

Jarzka committed:
Käytä oikeaa parametrin nimeä palvelujulkaisussa

Jarzka committed:
Korjaa SQL-syntaksivirhe

Jarzka committed:
Kytke tallennusnappi oikein tallennuspalveluun

Jarzka committed:
Toteuta palvelu laskentakohteiden haulle

Jarzka committed:
Hae urakan laskentakohteet frontille

Jarzka committed:
Korjaa buildi

Jarzka committed:
Lisää pari TODO-muistutusta

Jarzka committed:
Korjaa laskentakohteen valinnan kaatuminen

Jarzka committed:
Mahdollista kannasta tulleen laskentakohteen valinta lomakkeessa

Jarzka committed:
Alusta laskentakohde tyhjäksi uutta toteumaa tehtäessä

Jarzka committed:
Päivitä migraatio: laskentakohde ei ole pakollinen

Jarzka committed:
Ota laskentakohde mukaan toteuman tallennukseen

jarnovayrynen committed:
Pompauta migraationumeroa

jarnovayrynen committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into HAR-3682_yllapidon_muut_tyot

jarnovayrynen committed:
Lisää muokkausmetadata yllapito_toteumalle

jarnovayrynen committed:
Pompauta migraationumeroa

jarnovayrynen committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into HAR-3682_yllapidon_muut_tyot

jarnovayrynen committed:
Edistä laskentakohteen tallennusta

jarnovayrynen committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into HAR-3682_yllapidon_muut_tyot

jarnovayrynen committed:
Tallenna myös sopimusnumero muuhun työhön

jarnovayrynen committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into HAR-3682_yllapidon_muut_tyot

jarnovayrynen committed:
Hanskaa laskentakohde kun tallennetaan ylläpitototeuma

jarnovayrynen committed:
Lisää testi laskentakohteiden muuntamiselle

jarnovayrynen committed:
Merge branch 'develop' into HAR-3682_yllapidon_muut_tyot

jarnovayrynen committed:
Lisää testi ylläpidon toteumien haulle ja tallennukselle

jarnovayrynen committed:
Merge remote-tracking branch 'origin/HAR-3682_yllapidon_muut_tyot' into HAR-3682_yllapidon_muut_tyot

jarnovayrynen committed:
Lisää yllapitototeumatesti

jarnovayrynen committed:
Merge branch 'develop' into HAR-3682_yllapidon_muut_tyot

Jarzka commented on src/clj/harja/palvelin/palvelut/yllapito_toteumat.clj:
```diff
@@ -0,0 +1,107 @@
+(ns harja.palvelin.palvelut.yllapito_toteumat
```
Tässä pitäisi olla viiva, ei alaviiva

Jarzka commented on src/clj/harja/kyselyt/yllapito_toteumat.sql:
```diff
@@ -0,0 +1,56 @@
+  laskentakohde = :laskentakohde,
+  muokattu = NOW(),
+  muokkaaja = :kayttaja
+WHERE id = :id;
```
Tässä pitäisi olla WHERE-ehtona myös urakka-id, muuten vaarana, että päivitetään jonkin muun urakan toteuma. Palvelussa kyllä tarkistetaan, että käyttäjällä on kirjoitusoikeus antamaansa urakkaan, mutta toteuman id voi edelleen kuulua ihan muuhun urakkaan.

jarnovayrynen replied:
Tätä mietin itsekin, ja päätin poistaa urakka-id:n tarkistamisen redundanttina. Ainakin käyttöliittymässähän meillä on täysi kontrolli  että näkymässä ei ole muiden urakoiden toteumia. API:n kautta missä urakoitsijan järjestelmä voi tehdä virheen, voisi tuo urakka-id:nkin tarkistaminen olla perustelua. Voin lisätä sen takaisin kyllä, mutta jos tuota joskus tarvitaan, on meillä siinä tapauksessa muualla fataaleja bugeja.

Jarzka replied:
Frontissa tämä ei tosiaan ole ongelma, mutta ainahan joku voi yrittää lähettää pahan requestin, jossa id/urakka vaihdettu. Mielestäni olemme yleisesti pyrkineet siihen, että kun päivitetään urakkaan kuuluvia asioita, otetaa WHERE-ehtoon aina myös urakan id. :)

Jarzka commented on test/clj/harja/palvelin/palvelut/yllapito_toteumat_test.clj:
```diff
@@ -0,0 +1,102 @@
+(deftest tallenna-yllapito-toteuma-test
+         (let [urakka (hae-oulun-tiemerkintaurakan-id)
+               sopimus (hae-oulun-tiemerkintaurakan-paasopimuksen-id)
+               alkupvm (pvm/luo-pvm 2016 0 1)
```
Itse olisin halukas deprecoimaan tämän luo-pvm funktion, koska pitäisi pyrkiä käyttämään yhtä pvm-tyyppiä (Javan uudempaa date-APIa tai mieluummin joda-timeä, emme ole kai päässeet tästä yksimielisyyteen). luo-pvm käyttää vanhaa java.Datea.

Jarzka commented on src/cljs/harja/views/urakka/toteumat/tiemerkinta_muut_tyot.cljs:
```diff
@@ -0,0 +1,134 @@
+(defn- valinnat [e! {:keys [valittu-urakka valittu-sopimusnumero
+                            valitse-sopimusnumero valitun-urakan-hoitokaudet
+                            valittu-hoitokausi valitse-hoitokausi]}]
+  (let [muokkausoikeus? true] ;; TODO OIKEUSTARKISTUS
```
Täältä puuttuu oikeustarkistus

Jarzka commented on src/cljs/harja/views/urakka/toteumat/tiemerkinta_muut_tyot.cljs:
```diff
@@ -0,0 +1,134 @@
+(defn muu-tyo-lomake [e! tila {:keys [valittu-urakka valittu-sopimusnumero] :as riippuvuudet}]
+  (let [vanha-toteuma? (get-in tila [:valittu-toteuma :id])
+        valinnat (:valinnat tila)
+        muokkausoikeus? true] ;; TODO OIKEISTARKISTUS
```
Ja täältä myös

Jarzka commented on src/cljs/harja/views/urakka/toteumat/tiemerkinta_muut_tyot.cljs:
```diff
@@ -0,0 +1,134 @@
+      [{:otsikko "Päivämäärä" :nimi :pvm :tyyppi :pvm :pakollinen? true}
+       {:otsikko "Hinta" :nimi :hinta :tyyppi :positiivinen-numero :pakollinen? true}
+       {:otsikko "Ylläpitoluokka" :nimi :yllapitoluokka :tyyppi :positiivinen-numero
+        :huomauta [[:yllapitoluokka]]}
```
Sallii antaa ylläpitoluokaksi jotain muuta kuin 1-3, onko tarkoitus?

jarnovayrynen replied:
Ehkä sitä voidaan nyt tiukentaa. YHA API vaatii vain integerin, mutta vaaditaan me 1,2 tai 3 tai ei arvoa, ja hoidetaan sitä kautta oikea integer YHA API:in.

jarnovayrynen committed:
Pompauta migraationumeroa

jarnovayrynen committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into HAR-3682_yllapidon_muut_tyot

jarnovayrynen committed:
Korjaa ylläpitoluokan validointi: 1, 2, 3 tai nil

jarnovayrynen committed:
Varmista urakkakin kun muu työ päivitetään

jarnovayrynen committed:
Merge remote-tracking branch 'origin/HAR-3682_yllapidon_muut_tyot' into HAR-3682_yllapidon_muut_tyot

jarnovayrynen committed:
Uudelleennimeä yllapito-toteumat

jarnovayrynen committed:
Lisää oikeustarkistukset muut työt toteumiin

jarnovayrynen committed:
Järjestä muut työt pvm mukaan

jarnovayrynen committed:
Järjestä laskentakohteet aakkosjärjestykseen

Järjestä laskentakohteet aakkosjärjestykseen, paitsi 'Ei laskentakohdetta' oltava
aina ensimmäinen.

jarnovayrynen committed:
Järjestä muut työt -toteumat pvm mukaan jo kannasta

jarnovayrynen committed:
Korjaa crash kun :haetaan haun tuloksissa

jarnovayrynen committed:
Käytä Toteutus - Muut työt -oikeuksia

jarnovayrynen committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into HAR-3682_yllapidon_muut_tyot

jarnovayrynen committed:
Poista typo

jarnovayrynen committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into HAR-3682_yllapidon_muut_tyot

jarnovayrynen committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into HAR-3682_yllapidon_muut_tyot

jarnovayrynen committed:
Korjaa dev-restart TyokoneenPuhdistus

jarnovayrynen committed:
Korjaa tiemerkintäurakan oikeustarkistukset

jarnovayrynen committed:
Näytä talvihoitoluokka oikein tarkastuslistassa
